src/eula.rs

```diff
@@ -1,116 +1,272 @@
+
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use log::trace;
+use log::warn;
+
+use std::str::FromStr;
+
+use cargo_metadata::Package;
+
+/// License info
+#[derive(Clone, Debug)]
+pub struct Licenses {
+    /// The license for the actual source code
+    ///
+    /// This likely will become/contain a Vec at some point,
+    /// since dual MIT/Apache wants to have two license files!
+    pub source_license: Option<License>,
+    /// The end-user license (EULA) that must be agreed to when installing
+    pub end_user_license: Option<License>,
+}
+
+/// A license file/item
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct License {
+    /// Path to the file, relative to package's root dir.
+    ///
+    /// So generated files would be "wix\License.rtf"
+    /// and the typical LICENSE file would just be "LICENSE-MIT".
+    ///
+    /// Conveniently this means we don't need to do any special handling
+    /// to rewrite/relativize a path we get out of a Cargo.toml.
+    ///
+    /// It can also be an absolute path if the user passed this via CLI
+    /// and doesn't care about persistence/portability.
+    pub stored_path: StoredPathBuf,
+    /// File name to use for the license when installed to the user's system
+    ///
+    /// If None, the source file name is used.
+    pub name: Option<String>,
+
+    /// This file needs to be generated, write it to the given path
+    /// using the given Template.
+    pub generate: Option<(Utf8PathBuf, Template)>,
+}
+
+impl License {
+    /// Create a License entry with just the StoredPath
+    fn from_stored_path(path: &StoredPath) -> Self {
+        Self {
+            name: None,
+            stored_path: path.to_owned(),
+            generate: None,
+        }
+    }
+}
+
+impl Licenses {
+    /// Get license/eula info for a package
+    pub fn new(
+        dest_dir: Option<&Utf8Path>,
+        license_path: Option<&StoredPath>,
+        eula_path: Option<&StoredPath>,
+        package: &Package,
+    ) -> Result<Self> {
+        let source_license = Self::find_source_license(dest_dir, license_path, package)?;
+        let end_user_license =
+            Self::find_end_user_license(eula_path, package, source_license.as_ref())?;
+
+        Ok(Self {
+            source_license,
+            end_user_license,
+        })
+    }
+
+    /// Find the source-license for a package
+    fn find_source_license(
+        dest_dir: Option<&Utf8Path>,
+        path: Option<&StoredPath>,
+        package: &Package,
+    ) -> Result<Option<License>> {
+        trace!("finding source license for {}", package.name);
+        // If explicitly passed, use that
+        if let Some(path) = path {
+            trace!("explicit source-license path passed as argument, using that");
+            return Ok(Some(License::from_stored_path(path)));
+        }
+
+        let package_dir = package.manifest_path.parent().expect("non-root Cargo.toml");
+
+        // If there's [package.manifest.wix].license, respect that
```

Comment:
This is conflating the EULA and the sidecar License file, correct? Should there ultimately be a `eula` and `license` field? Maybe we actually need a hierarchy of some kind, not to over complicate, but a `license` field is the parent and if a `eula` and `sidecar` (or whatever) are defined, they override the `license` field, but if not, then the license is used for both the `eula` and `sidecar` and if the `license` is not defined, then auto-generated is used. Not sure about disabling one or both with a single field.

Reply:
Reading further down, there does appear to be a `eula` field separate from `license`, but I am still wondering about the hierarchy.

Reply:
The current hierarchy is that 

* we try all the different ways to find a source-license
* we try all the different ways to find an end-user-license
  * the last option being "use the source-license if it's an RTF (and eula=false wasn't set)"
  
I think it makes sense that they should be treated as separate/equals (as the code currently does). Personally I think the "use your source-license as a eula" thing is a bit of a silly cargo-cult, but as you say, people like doing it, so I'm happy to keep it working for people who like it (that said, cargo-dist will probably default to emitting eula=false/license=false into the user's Cargo.toml to make this stuff opt-in instead of opt-out).